pkg/quota/v1/evaluator/core/resource_claims.go

```diff
@@ -0,0 +1,166 @@
+func (p *claimEvaluator) Handles(a admission.Attributes) bool {
+	op := a.GetOperation()
+	if op == admission.Create {
+		return true
+	}
+	if op == admission.Update {
+		return true
+	}
+	return false
+}
+
+// Matches returns true if the evaluator matches the specified quota with the provided input item
+func (p *claimEvaluator) Matches(resourceQuota *corev1.ResourceQuota, item runtime.Object) (bool, error) {
+	return generic.Matches(resourceQuota, item, p.MatchingResources, generic.MatchesNoScopeFunc)
+}
+
+// MatchingScopes takes the input specified list of scopes and input object. Returns the set of scopes resource matches.
+func (p *claimEvaluator) MatchingScopes(item runtime.Object, scopes []corev1.ScopedResourceSelectorRequirement) ([]corev1.ScopedResourceSelectorRequirement, error) {
+	return []corev1.ScopedResourceSelectorRequirement{}, nil
+}
+
+// UncoveredQuotaScopes takes the input matched scopes which are limited by configuration and the matched quota scopes.
+// It returns the scopes which are in limited scopes but don't have a corresponding covering quota scope
+func (p *claimEvaluator) UncoveredQuotaScopes(limitedScopes []corev1.ScopedResourceSelectorRequirement, matchedQuotaScopes []corev1.ScopedResourceSelectorRequirement) ([]corev1.ScopedResourceSelectorRequirement, error) {
+	return []corev1.ScopedResourceSelectorRequirement{}, nil
+}
+
+// MatchingResources takes the input specified list of resources and returns the set of resources it matches.
+func (p *claimEvaluator) MatchingResources(items []corev1.ResourceName) []corev1.ResourceName {
+	result := []corev1.ResourceName{}
+	for _, item := range items {
+		if item == ClaimObjectCountName /* object count quota fields */ ||
+			strings.HasSuffix(string(item), corev1.ResourceClaimsPerClass /* by device class */) {
+			result = append(result, item)
+		}
+	}
+	return result
+}
+
+// Usage knows how to measure usage associated with item.
+func (p *claimEvaluator) Usage(item runtime.Object) (corev1.ResourceList, error) {
+	result := corev1.ResourceList{}
+	claim, err := toExternalResourceClaimOrError(item)
+	if err != nil {
+		return result, err
+	}
+
+	// charge for claim
+	result[ClaimObjectCountName] = *(resource.NewQuantity(1, resource.DecimalSI))
+	for _, request := range claim.Spec.Devices.Requests {
+		if request.AdminAccess {
+			// No need to count, admins get unlimited access.
+			continue
+		}
+		deviceClassClaim := V1ResourceByDeviceClass(request.DeviceClassName)
+		var numDevices int64
+		switch request.AllocationMode {
+		case resourceapi.DeviceAllocationModeExactCount:
+			numDevices = request.Count
+		case resourceapi.DeviceAllocationModeAll:
+			// Worst case...
+			numDevices = resourceapi.AllocationResultsMaxSize
+		default:
+			// Shouldn't happen because other modes don't pass validation.
+			// If a new mode gets added, this code needs to be updated.
+			return nil, fmt.Errorf("internal error: %s: unsupported count mode", request.AllocationMode)
```

Comment:
This stills vulnerable to skew.

Oops scenario
1. v1.40 adds new DeviceAllocationMode="Awesome"
2. Forget to update this.
3. insta-fail

Downgrade scenario
1. v1.40 adds value
2. downgrade to 1.39
3. fail until removing resourceclaim

Standard upgrade skew
1. add in v1.40
2. start upgrade, 1.40 kube-apiserver (moves first), 1.39 kube-contorller-manager (uses evaluators to refill quota)
3. fail

Reply:
Ack. So your previous proposal to simply not count such requests makes sense. It's consistent with "new alpha fields don't work after a downgrade".